### PR TITLE
Issue 15475: ddoc should not escape unbalanced parens in code blocks twice.

### PR DESCRIPTION
**Pull request by quickfur**
Basically, the root cause of this bug is that unbalanced parens inside code block comments are escaped twice: once by `writeSections`, the second time by `highlightCode2`. The first time it gets escaped to `$(RPAREN)`, the second time it gets escaped to `$(DOLLAR)$(LPAREN)RPAREN$(RPAREN)`. But since macros are only expanded once, the output gets stuck with a literal `$(RPAREN)` instead of the original `)`. So one of these escapes must be suppressed.

However, we can't suppress this in `highlightCode2`, because it gets called via other code paths where the input may not have been escaped yet, so it will break other cases. So that leaves us with modifying `escapeStrayParenthesis` to ignore code blocks, since code blocks eventually end up in one of `highlightCode` or `highlightCode2`, which already do any required escaping anyway.

### CODE REVIEW

File: src/doc.d [hunk 1]
```diff
@@ -657,25 +657,30 @@ extern (C++) void escapeDdocString(OutBuffer* buf, size_t start)
 extern (C++) void escapeStrayParenthesis(Loc loc, OutBuffer* buf, size_t start)
 {
     uint par_open = 0;
+    int inCode = 0;
```
Comment (braddr) [inline src/doc.d]: should be bool

File: src/doc.d [hunk 2]
```diff
@@ -686,6 +691,18 @@ extern (C++) void escapeStrayParenthesis(Loc loc, OutBuffer* buf, size_t start)
...
+            // Issue 15465: don't try to escape unbalanced parens inside code
+            // blocks.
+            int numdash = 0;
+            while (u < buf.offset && buf.data[u] == '-')
+            {
+                numdash++;
+                u++;
+            }
+            if (numdash >= 3)
+                inCode ^= 1;
```
Comment (braddr) [inline src/doc.d]: inCode = !inCode;

File: src/doc.d (hunk 1 above)
Comment (quickfur) [inline src/doc.d]: Fixed.

**Comment by braddr** [comment]
For even more correctness / error checking, it might be good to have code blocks essentially maintain a stack of parenthesis, so that on exit of a code block, the value it had on entry is restored.

File: src/doc.d (hunk 2 above)
Comment (quickfur) [inline src/doc.d]: Fixed.

**Comment by quickfur** [comment]
I wonder if now would be a good opportunity to refactor the comment-parsing code into some kind of range-based API that can be used for keeping track of parens nesting, code blocks, etc., since this is used in multiple places, and right now it's just copy-pasta everywhere.

**Comment by quickfur** [comment]
Though I was thinking it might be better to keep this PR small and focused for now, and worry about refactoring in a follow-up PR.

A lot of code in `doc.d` is begging to be refactored.

**Comment by braddr** [comment]
squash please?  Then I'm good with this change.  I agree that the code and behavior as is need work.  This change is at best a bandaid for a single specific problem.

**Comment by quickfur** [comment]
Squashed.

**Comment by braddr** [comment]
Auto-merge toggled on

**Pull request merged**